### PR TITLE
Starting testing in Rails

### PR DESCRIPTION
**Pull request by C4L86**

### CODE REVIEW

File: app/helpers/application_helper.rb [hunk 1]
```diff
@@ -1,59 +1,67 @@
...
-    link_to(album.title, "/albums/#{album.id}")
+    link_to(album.title, "/albums/#{album.id}", {class: "show_link"})
   end
   
   def albums_show_image_link(album)
-    link_to(image_tag(album.album_images.cover_image, size:"100x100"), "/albums/#{album.id}")
+    link_to(image_tag(album.album_images.cover_image, size:"200x200"), "/albums/#{album.id}", {class: "show_album_image_link"})
+  end
+
+  def albums_user_addition_link
```
Comment (sumeetjain) [inline app/helpers/application_helper.rb:15]: Remove unused methods (applicationhelper#albums_user_addition_link)

File: app/helpers/application_helper.rb [hunk 2]
```diff
@@ -1,59 +1,67 @@
...
-    link_to("Genres", "/genres/index")
+    link_to("Genres", "/genres/index", {class: "index_link"})
   end
 
   def genres_show_link(genre)
-    link_to(genre.name, "/genres/#{genre.id}")
+    link_to(genre.name, "/genres/#{genre.id}", {class: "show_link"})
+  end
+
+  def hear_nebraska_link
```
Comment (sumeetjain) [inline app/helpers/application_helper.rb:55]: Remove unused methods (applicationhelper#hear_nebraska_link)

File: app/helpers/application_helper.rb [hunk 3]
```diff
@@ -1,59 +1,67 @@
 module ApplicationHelper
   
   def albums_index_link
-    link_to("Albums", "/albums/index")
+    link_to("The Vinyl Archive", "/albums/index", {class: "vinyl_archive"})
   end
 
   def albums_show_link(album)
-    link_to(album.title, "/albums/#{album.id}")
+    link_to(album.title, "/albums/#{album.id}", {class: "show_link"})
```
Comment (sumeetjain) [inline app/helpers/application_helper.rb:8]: Refers to album more than self (maybe move it to another class?) (FeatureEnvy)

File: app/helpers/application_helper.rb [hunk 4]
```diff
@@ -1,59 +1,67 @@
...
   end
 
   def areas_index_link
-    link_to("Areas", "/areas/index")
+    link_to("Cities", "/areas/index", {class: "index_link"})
   end
 
   def areas_show_link(area)
-    link_to(area.name, "/areas/#{area.id}")
+    link_to(area.name, "/areas/#{area.id}", {class: "show_link"})
```
Comment (sumeetjain) [inline app/helpers/application_helper.rb:36]: Refers to area more than self (maybe move it to another class?) (FeatureEnvy)

File: app/helpers/application_helper.rb [hunk 5]
```diff
@@ -1,59 +1,67 @@
...
   end
 
   def artists_index_link
-    link_to("Artists", "artists/index")
+    link_to("Artists", "artists/index", {class: "index_link"})
   end
 
   def artists_show_link(artist)
-    link_to(artist.name, "/artists/#{artist.id}")
+    link_to(artist.name, "/artists/#{artist.id}", {class: "show_link"})
```
Comment (sumeetjain) [inline app/helpers/application_helper.rb:24]: Refers to artist more than self (maybe move it to another class?) (FeatureEnvy)

File: app/helpers/application_helper.rb [hunk 6]
```diff
@@ -1,59 +1,67 @@
...
   end
 
   def formats_index_link
-    link_to("Formats", "/formats/index")
+    link_to("Formats", "/formats/index", {class: "index_link"})
   end
 
   def formats_show_link(format)
-    link_to(format.name, "/formats/#{format.id}")
+    link_to(format.name, "/formats/#{format.id}", {class: "show_link"})
```
Comment (sumeetjain) [inline app/helpers/application_helper.rb:64]: Refers to format more than self (maybe move it to another class?) (FeatureEnvy)

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -1,59 +1,67 @@
...
   end
 
   def genres_index_link
-    link_to("Genres", "/genres/index")
+    link_to("Genres", "/genres/index", {class: "index_link"})
   end
 
   def genres_show_link(genre)
-    link_to(genre.name, "/genres/#{genre.id}")
+    link_to(genre.name, "/genres/#{genre.id}", {class: "show_link"})
```
Comment (sumeetjain) [inline app/helpers/application_helper.rb:52]: Refers to genre more than self (maybe move it to another class?) (FeatureEnvy)

File: app/helpers/application_helper.rb [hunk 8]
```diff
@@ -1,59 +1,67 @@
...
   end
 
   def albums_show_link(album)
-    link_to(album.title, "/albums/#{album.id}")
+    link_to(album.title, "/albums/#{album.id}", {class: "show_link"})
   end
   
   def albums_show_image_link(album)
-    link_to(image_tag(album.album_images.cover_image, size:"100x100"), "/albums/#{album.id}")
+    link_to(image_tag(album.album_images.cover_image, size:"200x200"), "/albums/#{album.id}", {class: "show_album_image_link"})
```
Comment (sumeetjain) [inline app/helpers/application_helper.rb:12]: Space missing after colon.

Comment (sumeetjain) [inline app/helpers/application_helper.rb:12]: Line is too long. [127/80]

File: app/helpers/application_helper.rb [hunk 9]
```diff
@@ -1,59 +1,67 @@
...
+  end
+
+  def albums_user_addition_link
+    link_to("Add an Album", "/albums/user_addition", {class: "albums_user_addition"})
```
Comment (sumeetjain) [inline app/helpers/application_helper.rb]: Line is too long. [85/80]

File: app/helpers/application_helper.rb [hunk 10]
```diff
@@ -1,59 +1,67 @@
...
   end
 
   def artists_show_link(artist)
-    link_to(artist.name, "/artists/#{artist.id}")
+    link_to(artist.name, "/artists/#{artist.id}", {class: "show_link"})
   end
   
   def artists_show_image_link(artist)
-    link_to(image_tag(artist.artist_images.image_2, size:"100x100"), "/artists/#{artist.id}")
+    link_to(image_tag(artist.artist_images.image_2, size:"100x100"), "/artists/#{artist.id}", {class: "artists_show_image_link"})
```
Comment (sumeetjain) [inline app/helpers/application_helper.rb:28]: Space missing after colon.

Comment (sumeetjain) [inline app/helpers/application_helper.rb:28]: Line is too long. [129/80]

File: app/helpers/application_helper.rb [hunk 11]
```diff
@@ -1,59 +1,67 @@
...
+    link_to("Genres", "/genres/index", {class: "index_link"})
   end
 
   def genres_show_link(genre)
-    link_to(genre.name, "/genres/#{genre.id}")
+    link_to(genre.name, "/genres/#{genre.id}", {class: "show_link"})
+  end
+
+  def hear_nebraska_link
+    link_to("by Hear Nebraska", "https://www.hearnebraska.org", {class:"hear_nebraska"})
```
Comment (sumeetjain) [inline app/helpers/application_helper.rb:56]: Space missing after colon.

Comment (sumeetjain) [inline app/helpers/application_helper.rb:56]: Line is too long. [88/80]

File: config/routes.rb [hunk 12]
```diff
@@ -30,8 +30,10 @@
   get "albums/new"     => 'albums#new'
   post "albums/create" => 'albums#create'
   get "albums/:id/edit" => 'albums#edit'
+  # Is there a reason why this is a 'put' request?
   put "albums/:id/update" => 'albums#update'
   delete "albums/:id/delete" => 'albums#delete'
+
```
Comment (sumeetjain) [inline config/routes.rb:36]: Trailing whitespace detected.

File: config/routes.rb [hunk 13]
```diff
@@ -39,6 +41,7 @@
   get "artists/:id/new_location_and_image" => 'artists#new_location_and_image'
   post "artist/:id/create_members" => 'artists#create_members'
   get "artists/:id/new_members" => 'artists#new_members'
+
```
Comment (sumeetjain) [inline config/routes.rb:44]: Trailing whitespace detected.

File: app/models/album.rb [hunk 14]
```diff
@@ -26,19 +27,30 @@ def tracklist
   end
 
   # Returns the release date for a given album from the associated releases table
+  # I BELIEVE there is a way to achieve this with ActiveRecord relations
   def show_release_date
-    Release.find_by_id(self.id).release_date
+    release = Release.where({"album_id" => self.id})
+    release.each do |release|
```
Comment (sumeetjain) [inline app/models/album.rb]: Shadowing outer local variable - `release`.

File: app/models/album.rb [hunk 15]
```diff
@@ -26,19 +27,30 @@ def tracklist
...
+    end
+    @date
   end
 
   # Returns the label for a given album from the associated releases table
+  # I BELIEVE there is a way to achieve this with ActiveRecord relations
   def show_album_label
-    Release.find_by_id(self.id).label
+    release = Release.where({"album_id" => self.id})
+    release.each do |release|
```
Comment (sumeetjain) [inline app/models/album.rb]: Shadowing outer local variable - `release`.

File: app/models/album.rb [hunk 16]
```diff
@@ -26,19 +27,34 @@ def tracklist
   end
 
   # Returns the release date for a given album from the associated releases table
+  # I BELIEVE there is a way to achieve this with ActiveRecord relations
   def show_release_date
-    Release.find_by_id(self.id).release_date
+    @date = []
+    release = []
```
Comment (sumeetjain) [inline app/models/album.rb]: Useless assignment to variable - `release`.

File: app/models/album.rb [hunk 17]
```diff
@@ -26,19 +27,34 @@ def tracklist
...
+    release = Release.where({"album_id" => self.id})
+    release.each do |release|
```
Comment (sumeetjain) [inline app/models/album.rb]: Shadowing outer local variable - `release`.

File: app/models/album.rb [hunk 18]
```diff
@@ -26,19 +27,34 @@ def tracklist
...
+    end
+    @date
   end
 
   # Returns the label for a given album from the associated releases table
+  # I BELIEVE there is a way to achieve this with ActiveRecord relations
   def show_album_label
-    Release.find_by_id(self.id).label
+    @label = []
+    release = []
```
Comment (sumeetjain) [inline app/models/album.rb]: Useless assignment to variable - `release`.

File: app/models/album.rb [hunk 19]
```diff
@@ -26,19 +27,34 @@ def tracklist
...
+    release = Release.where({"album_id" => self.id})
+    release.each do |release|
```
Comment (sumeetjain) [inline app/models/album.rb]: Shadowing outer local variable - `release`.

File: app/models/album.rb [hunk 20]
```diff
@@ -26,19 +27,47 @@ def tracklist
...
-    Format.find_by_id(format)
+    @formats = []
+    @format_ids = []
+    releases = []
+
+    releases = Release.where({"album_id" => self.id})
+    releases.each do |release|
+      @format_ids << release.format_id
+    end
+
```
Comment (sumeetjain) [inline app/models/album.rb:76]: Remove trailing whitespace

File: app/models/album.rb [hunk 21]
```diff
@@ -26,19 +27,47 @@ def tracklist
   end
 
   # Returns the release date for a given album from the associated releases table
+  # I BELIEVE there is a way to achieve this with ActiveRecord relations
   def show_release_date
-    Release.find_by_id(self.id).release_date
+    @dates = []
+    releases = []
```
Comment (sumeetjain) [inline app/models/album.rb]: Useless assignment to variable - `releases`.

File: app/models/album.rb [hunk 22]
```diff
@@ -26,19 +27,47 @@ def tracklist
...
+    end
+    @dates
   end
 
   # Returns the label for a given album from the associated releases table
+  # I BELIEVE there is a way to achieve this with ActiveRecord relations
   def show_album_label
-    Release.find_by_id(self.id).label
+    @labels = []
+    releases = []
```
Comment (sumeetjain) [inline app/models/album.rb]: Useless assignment to variable - `releases`.

File: app/models/album.rb [hunk 23]
```diff
@@ -26,19 +27,47 @@ def tracklist
...
 
   # Retrieves the format_id for a given album from the associated releases table, then returns the format for that album
   # Refactored by removing the '.name' call on at the end of the last line to accommodate the 'formats_show_link' method
+  # I BELIEVE there is a way to achieve this with ActiveRecord relations
   def show_album_format
-    format = Release.find_by_id(self.id).format_id
-    Format.find_by_id(format)
+    @formats = []
+    @format_ids = []
+    releases = []
```
Comment (sumeetjain) [inline app/models/album.rb]: Useless assignment to variable - `releases`.

File: app/models/album.rb (hunk 20 above)
Comment (sumeetjain) [inline app/models/album.rb:76]: Trailing whitespace detected.

File: app/views/albums/show.html.erb [hunk 24]
```diff
@@ -19,15 +19,33 @@
 <br>
 
 <!-- Shows album's release date -->
-<p><b>Released:</b> <%= @album.show_release_date %> </p>
+<p>
+<b>Released:</b>
```
Comment (sumeetjain) [inline app/views/albums/show.html.erb:23]: Remove trailing whitespace

File: app/models/album.rb [hunk 25]
```diff
@@ -25,20 +26,47 @@ def tracklist
     tracklist
   end
 
-  # Returns the release date for a given album from the associated releases table
+  # Returns hash of release dates for a given album from the associated releases table
```
Comment (sumeetjain) [inline app/models/album.rb:45]: Line is too long. [86/80]

File: app/models/album.rb [hunk 26]
```diff
@@ -25,20 +26,46 @@ def tracklist
...
   def show_album_format
-    format = Release.find_by_id(self.id).format_id
-    Format.find_by_id(format)
+    formats = []
+    format_ids = []
+
+    releases.each do |release|
+      format_ids << release.format_id
+    end
+
```
Comment (sumeetjain) [inline app/models/album.rb:76]: Remove trailing whitespace

File: app/models/album.rb [hunk 27]
```diff
@@ -25,20 +26,46 @@ def tracklist
...
-  # Refactored by removing the '.name' call on at the end of the last line to accommodate the 'formats_show_link' method
+  # I BELIEVE there is a way to achieve this with ActiveRecord relations
   def show_album_format
-    format = Release.find_by_id(self.id).format_id
-    Format.find_by_id(format)
+    formats = []
+    format_ids = []
+
+    releases.each do |release|
+      format_ids << release.format_id
```
Comment (sumeetjain) [inline app/models/album.rb:74]: Refers to format_ids more than self (maybe move it to another class?) (FeatureEnvy)

File: app/models/album.rb [hunk 28]
```diff
@@ -25,20 +26,46 @@ def tracklist
     tracklist
   end
 
-  # Returns the release date for a given album from the associated releases table
+  def releases
+    releases = []
```
Comment (sumeetjain) [inline app/models/album.rb]: Useless assignment to variable - `releases`.

File: app/models/album.rb [hunk 29]
```diff
@@ -25,20 +26,46 @@ def tracklist
...
+    releases = Release.where({"album_id" => self.id})
```
Comment (sumeetjain) [inline app/models/album.rb:31]: Useless assignment to variable - `releases`.

File: app/models/album.rb [hunk 30]
```diff
@@ -25,20 +26,46 @@ def tracklist
...
+  end
+
+  # Returns hash of release dates for a given album from the associated releases table
```
Comment (sumeetjain) [inline app/models/album.rb:45]: Line is too long. [86/80]

File: app/models/album.rb (hunk 26 above)
Comment (sumeetjain) [inline app/models/album.rb:76]: Trailing whitespace detected.

File: app/models/album.rb [hunk 31]
```diff
@@ -25,20 +26,55 @@ def tracklist
     tracklist
   end
 
-  # Returns the release date for a given album from the associated releases table
+  def album_releases
+    releases = Release.where({"album_id" => self.id})
+  end
+
+  # Attempting to refactor the below 'show' methods into a master method
+  def show_release(attribute)
```
Comment (sumeetjain) [inline app/models/album.rb]: Remove unused methods (album#show_release)

File: app/models/album.rb [hunk 32]
```diff
@@ -25,20 +26,55 @@ def tracklist
...
   def show_album_format
-    format = Release.find_by_id(self.id).format_id
-    Format.find_by_id(format)
+    formats = []
+    format_ids = []
+
+    album_releases.each do |release|
+      format_ids << release.format_id
+    end
+
```
Comment (sumeetjain) [inline app/models/album.rb:76]: Remove trailing whitespace

File: app/models/album.rb [hunk 33]
```diff
@@ -25,20 +26,55 @@ def tracklist
...
-  # Refactored by removing the '.name' call on at the end of the last line to accommodate the 'formats_show_link' method
+  # I BELIEVE there is a way to achieve this with ActiveRecord relations
   def show_album_format
-    format = Release.find_by_id(self.id).format_id
-    Format.find_by_id(format)
+    formats = []
+    format_ids = []
+
+    album_releases.each do |release|
+      format_ids << release.format_id
```
Comment (sumeetjain) [inline app/models/album.rb:74]: Refers to format_ids more than self (maybe move it to another class?) (FeatureEnvy)

File: app/models/album.rb (hunk 31 above)
Comment (sumeetjain) [inline app/models/album.rb]: Has unused parameter 'attribute' (UnusedParameters)

Comment (sumeetjain) [inline app/models/album.rb]: Unused method argument - `attribute`. If it's necessary, use `_` or `_attribute` as an argument name to indicate that it won't be used. You can also write as `show_release(*)` if you want the method to accept any arguments but don't care about them.

File: app/models/album.rb [hunk 34]
```diff
@@ -25,20 +26,55 @@ def tracklist
...
+  def show_release(attribute)
+    array = []
+
+    album_releases.each do |release|
+      array << release.attribute
+    end
+    array
+  end
+
+  # Returns hash of release dates for a given album from the associated releases table
```
Comment (sumeetjain) [inline app/models/album.rb:45]: Line is too long. [86/80]

File: app/models/album.rb (hunk 32 above)
Comment (sumeetjain) [inline app/models/album.rb:76]: Trailing whitespace detected.

File: app/models/album.rb [hunk 35]
```diff
@@ -25,20 +26,55 @@ def tracklist
     tracklist
   end
 
-  # Returns the release date for a given album from the associated releases table
+  def album_releases
+    releases = Release.where({"album_id" => self.id})
+  end
+
+  # Attempting to refactor the below 'show' methods into a master method
+  def show_album(attribute)
```
Comment (sumeetjain) [inline app/models/album.rb:36]: Remove unused methods (album#show_album)

File: app/models/album.rb [hunk 36]
```diff
@@ -25,20 +26,55 @@ def tracklist
...
+    array = []
+
```
Comment (sumeetjain) [inline app/models/album.rb:38]: Remove trailing whitespace

File: app/models/album.rb (hunk 35 above)
Comment (sumeetjain) [inline app/models/album.rb:36]: Unused method argument - `attribute`. If it's necessary, use `_` or `_attribute` as an argument name to indicate that it won't be used. You can also write as `show_album(*)` if you want the method to accept any arguments but don't care about them.

File: app/models/album.rb (hunk 36 above)
Comment (sumeetjain) [inline app/models/album.rb:38]: Trailing whitespace detected.

File: app/models/album.rb [hunk 37]
```diff
@@ -25,20 +26,56 @@ def tracklist
...
   end
 
-  # Returns the release date for a given album from the associated releases table
+  # Extracted method for below methods
+  def album_releases
+    releases = Release.where({"album_id" => self.id})
+  end
+
+  # Attempting to refactor the below 'show' methods into a master method
+  def show_album(attribute)
```
Comment (sumeetjain) [inline app/models/album.rb:36]: Has unused parameter 'attribute' (UnusedParameters)

**Comment by C4L86** [comment]
I have been struggling with this error since yesterday and was hoping to get another set of eyes on it. Currently when trying to access the albums/index view, this is the error that appears:
screen shot 2016-04-04 at 12 08 47 pm
However, when in a binding.pry it's showing that `album` is not `nil`:
screen shot 2016-04-04 at 12 21 13 pm
 The controller actions and routes all look good so I am at a loss as to why this error is occurring.

**Comment by halfghaninne** [comment]
Do you have any albums with no corresponding `album_image` record? That error indicates to me that for some album, `.album_images` is nil, and therefor `.cover_image` can't be called.

**Comment by C4L86** [comment]
In my local DB, I have cover images for all albums, and each image in question is called just fine on the other pages. Could I come up and chat for a minute if you're free?

**Pull request merged**